File: src/test/java/com/backend/immilog/global/presentation/controller/ImageControllerTest.java
package com.backend.immilog.global.presentation.controller;

import com.backend.immilog.image.application.service.ImageService;
import com.backend.immilog.image.domain.enums.ImageType;
import com.backend.immilog.image.presentation.request.ImageRequest;
import com.backend.immilog.image.presentation.controller.ImageController;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.http.ResponseEntity;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;
import java.util.Objects;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.*;
import static org.springframework.http.HttpStatus.NO_CONTENT;
import static org.springframework.http.HttpStatus.OK;

@DisplayName("이미지 컨트롤러 테스트")
class ImageControllerTest {
    private final ImageService imageService = mock(ImageService.class);
    private final ImageController imageController = new ImageController(imageService);

    @Test
    @DisplayName("이미지 업로드")
    void uploadImage() {
        // given
        List<MultipartFile> files = List.of(mock(MultipartFile.class));
        String imagePath = "imagePath";
        List<String> imageDTO = List.of("imageUrl");
        when(imageService.saveFiles(files, imagePath, ImageType.POST))
                .thenReturn(imageDTO);
        // when
        ResponseEntity<ImageController.ImageResponse> response =
                imageController.uploadImage(files, imagePath, ImageType.POST);
        // then
        assertThat(response.getStatusCode()).isEqualTo(OK);
        List<String> data = (List<String>) Objects.requireNonNull(response.getBody()).data();
        assertThat(data.get(0)).isEqualTo("imageUrl");
    }

    @Test
    @DisplayName("이미지 삭제")
    void deleteImage() {
        // given
        String imagePath = "imagePath";
        ImageRequest param = ImageRequest.builder()
                .imageDirectory("directory")
                .imagePath(imagePath)
                .build();
        // when
        ResponseEntity<ImageController.ImageResponse> response = imageController.deleteImage(param);
        // then
        verify(imageService, times(1)).deleteFile(imagePath);
        assertThat(response.getStatusCode()).isEqualTo(NO_CONTENT);
    }
}

File: src/test/java/com/backend/immilog/image/ImageControllerTest.kt
package com.backend.immilog.image

import com.backend.immilog.image.application.ImageUploadUseCase
import com.backend.immilog.image.domain.ImageType
import com.backend.immilog.image.presentation.ImageController
import com.backend.immilog.image.presentation.ImageRequest
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.mockito.Mockito.*
import org.springframework.http.HttpStatus.NO_CONTENT
import org.springframework.http.HttpStatus.OK
import org.springframework.web.multipart.MultipartFile

@DisplayName("이미지 컨트롤러 테스트")
class ImageControllerTest {
    private val imageUploader: ImageUploadUseCase = mock(ImageUploadUseCase.ImageUploader::class.java)
    private val imageController = ImageController(imageUploader)

    @Test
    @DisplayName("이미지 업로드")
    fun uploadImage() {
        // given
        val files: List<MultipartFile> = listOf(mock(MultipartFile::class.java))
        val imagePath = "imagePath"
        val imageDTO: List<String> = listOf("imageUrl")
        `when`(imageUploader.saveFiles(files, imagePath, ImageType.POST)).thenReturn(imageDTO)
        // when
        val response = imageController.uploadImage(files, imagePath, ImageType.POST)
        // then
        assertThat(response.statusCode).isEqualTo(OK)
        val data: List<String> = response.body?.data ?: emptyList()
        assertThat(data.first()).isEqualTo("imageUrl")
    }

    @Test
    @DisplayName("이미지 삭제")
    fun deleteImage() {
        // given
        val imagePath = "imagePath"
        val param = ImageRequest("directory", imagePath)
        // when
        val response = imageController.deleteImage(param)
        // then
        verify(imageUploader, times(1)).deleteFile(imagePath,"")
        assertThat(response.statusCode).isEqualTo(NO_CONTENT)
    }
}
